### PR TITLE
Add checked arithmetic v2

### PR DESCRIPTION
**Pull request by pvdrz**
This is a retake on #96 

cc @DarkDrek

### CODE REVIEW

File: src/machine/builder.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        Machine {
+            env: self.env,
+            arithmetic: self.arithmetic,
+        }
+    }
+
+    pub fn with_env<W2: Write>(self, env: Env<W2>) -> MachineBuilder<W2, A> {
+        MachineBuilder {
+            env,
+            arithmetic: self.arithmetic,
```
Comment (DarkDrek) [inline src/machine/builder.rs:34]: Does `..self` work here? Then we wouldn't need to change these methods when we decide to add more fields.

File: tests/mod.rs [hunk 2]
```diff
@@ -2,12 +2,25 @@ extern crate pijama;
...
+
 mod ast;
 mod eval;
 mod parse;
 mod type_check;
 mod util;
 
+fn run(input: &str) -> LangResult<Term> {
+    let machine = MachineBuilder::default()
+        .with_arithmetic(CheckedArithmetic)
```
Comment (DarkDrek) [inline tests/mod.rs]: Can we extract the default `MachineBuilder` to use in tests in its own function? So if we need some additional settings in some test we don't need to repeat all other options.

File: src/options.rs [hunk 3]
```diff
@@ -5,4 +5,15 @@ use structopt::StructOpt;
...
+    #[structopt(flatten)]
+    pub machine_opts: MachineOptions,
+}
+
+#[derive(Debug, StructOpt)]
+pub struct MachineOptions {
+    #[structopt(
+        long = "--overflow-check",
+        help = "Execution panics on integer overflow"
+    )]
```
Comment (DarkDrek) [inline src/options.rs:17]: Like in the old version of this PR I would say what is the default value or make it a required option.

Comment (pvdrz) [reply src/options.rs:17]: Oh sure I forgot!

**Comment by DarkDrek** [comment]
Can you add some tests that checks for the panics?

File: src/machine/builder.rs (hunk 1 above)
Comment (pvdrz) [reply src/machine/builder.rs:34]: sadly no because the new builder and the old one have different types :/

**Comment by pvdrz** [comment]
@DarkDrek do you think this is ready?

File: src/machine/arithmetic.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
+use std::convert::TryFrom;
+
+use pijama_ast::{BinOp, BinOp::*, UnOp, UnOp::*};
+
+/// Trait determining how arithmetic operations should be handled.
+pub trait Arithmetic {
+    fn native_bin_op(op: BinOp, n1: i64, n2: i64) -> i64;
+    fn native_un_op(op: UnOp, n: i64) -> i64;
```
Comment (DarkDrek) [inline src/machine/arithmetic.rs]: Do we want to keep these names? I can't find better name right now.

Comment (pvdrz) [reply src/machine/arithmetic.rs]: we can drop the `native_`. What do you think?

**Comment by DarkDrek** [comment]
> @DarkDrek do you think this is ready?

I'm fine with the direction of the changes but maybe we/you 🙏 can find a better name for methods in `Arithmetic`

File: src/machine/arithmetic.rs (hunk 4 above)
Comment (DarkDrek) [reply src/machine/arithmetic.rs]: Do we want to use the full name `binary_operation` and `unary_operation`?

Comment (pvdrz) [reply src/machine/arithmetic.rs]: SGTM

**Comment by DarkDrek** [comment]
Go ahead and merge after the checks. 👍

**Pull request merged**